### PR TITLE
prompt, save and restore password with libsecret

### LINKED ISSUE #20: Add gnome-keyring integration
Using session-wide secret management via the gnome-keyring service it would be possible for users to handle their CardDav credentials in a similar way to other remote credentials they require, including dealing with passphrase unlocking. This may also reduce the access time on `mcds` invocation as it would avoid fork-exec of `gpg` to access credentials.

On the downside it would mean linking with a whole load of additional dependencies and using the `dbus`.

### PR DESCRIPTION
**Pull request by andy-bower**
This attempt to close #20 was surprisingly straightforward. Some caveats come to mind:

1. I am new to autoconf so the `configure.ac` changes are a bit cargo-culty - may need some attention!
2. The ordering for trying credentials needs review: store -> gpg -> netrc -> prompt.
3. Do you want the password prompt at all when the store is not available?
4. Should it be possible to disable the store at runtime?
5. The `options.pwprompt` setting gets overwritten with whether the prompt was actually used so that _main_ knows whether to request the password be saved upon a successful query (we don't save the decrypted or netrc passsword!). That's a change of semantics versus the other 'options'.

This is what the store looks like:
```
$ lssecret
Item:	Mutt CardDAV Search user credentials
Key:	url
Value:	https://my.cloud.server/remote.php/dav/addressbooks/users/fred/contacts
Key:	user
Value:	fred
Key:	xdg:schema
Value:	com.github.t-brown.mcds.password
```

How to make a first query
```
$ src/mcds -u http://foobar/ bloggs
mcds: password for fred at http://foobar/:
```

The password then gets stored if successful and the prompt not presented again unless run with `-p`.

Tested with `libsecret-1-dev` not present and with `./configure --disable-libsecret`.

```
$ src/mcds -V
mcds (GNU mcds) 1.7
Copyright (C) 2019 Timothy Brown.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Compiled on Oct 21 2024 at 22:17:16:
  with GPGME support and
  with libsecret support.
```

### CODE REVIEW

**Comment by andy-bower** [comment]
I've rebased, squashed and force-pushed this change and will switch from draft for proper review.

File: src/rc.c [hunk 1]
```diff
@@ -178,7 +179,19 @@ read_rc(const char *file)
 		abs_file = NULL;
 	}
 
-	if (pfile) {
+	if (options.username == NULL) {
+		options.username = getenv("USER");
```
Comment (andy-bower) [inline src/rc.c]: Needs to be `strdup()`ped because this gets `free()`d later.

File: src/main.c [hunk 2]
```diff
@@ -117,15 +117,17 @@ main(int argc, char **argv)
...
-		fprintf(stderr, "  Use .netrc : %d\n", options.netrc);
-		fprintf(stderr, "  Username   : %s\n", options.username);
-		fprintf(stderr, "  Password   : %s\n", options.password);
-		fprintf(stderr, "  Query term : %s\n", options.term);
-		fprintf(stderr, "  Query      : %s\n",
+		fprintf(stderr, "  URL               : %s\n", options.url);
+		fprintf(stderr, "  SSL Verify        : %d\n", options.verify);
+		fprintf(stderr, "  Use .netrc        : %d\n", options.netrc);
+		fprintf(stderr, "  Use libsecret     : %d\n", options.libsecret);
+		fprintf(stderr, "  Password prompted : %d\n", options.pwprompt);
```
Comment (andy-bower) [inline src/main.c:128]: The reason I had "prompted" before was that I was modifying the options value with the result of whether it was actually prompted, for the purpose of saving it. If that is no longer the case you could reword it to make more sense as a config key, e.g. just "prompt".

File: src/main.c [hunk 3]
```diff
@@ -308,8 +323,12 @@ Copyright (C) %s Timothy Brown.\n\
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
 This is free software: you are free to change and redistribute it.\n\
 There is NO WARRANTY, to the extent permitted by law.\n\n"), "2019");
-	printf(_("Compiled on %s at %s %s GPGME support.\n\n"),
-	       __DATE__, __TIME__, ngettext("with", "with-out", HAVE_GPGME)
+	printf(_("Compiled on %s at %s:\n"
+		 " -  %s GPGME support.\n"
+		 " - %s libsecret support.\n\n"),
```
Comment (andy-bower) [inline src/main.c]: inconsistent spacing between those two lines?

File: src/secret.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+#include "defs.h"
+#include "mem.h"
+#include "options.h"
+#include "prompt.h"
+#include "secret.h"
+
+#if HAVE_LIBSECRET
+#include <libsecret/secret.h>
+
+#define MCDS_NAMESPACE "com.github.t-brown.mcds"
```
Comment (t-brown) [inline src/secret.c:52]: I'm not crash hot on `#define`s and prefer `static const char`. However, this does mean you can concat them easily.

File: src/rc.c [hunk 5]
```diff
@@ -178,7 +186,27 @@ read_rc(const char *file)
...
 	}
 
-	if (pfile) {
+	if (options.username == NULL) {
+		options.username = strdup(getenv("USER"));
+	}
+
+	if (options.pwprompt) {
+		prompt_password();
+	}
```
Comment (andy-bower) [inline src/rc.c:195]: Could we not also prompt for the password if `libsecret` is specified but a password is not found @t-brown? That would be how an application would normally behave - see if we've saved the password already and prompt the user if not found?

(That's why I originally described the option as 'forcing' a prompt because a prompt could also appear without it and the option could be used to clear the password, which admittedly is not a normal function of an application but was handy for testing!)

**Review by t-brown — APPROVED** [review]
Thanks for the contribution!

File: src/rc.c [hunk 6]
```diff
@@ -178,7 +186,27 @@ read_rc(const char *file)
...
+
+#if HAVE_LIBSECRET
+	if (options.libsecret) {
+		if (!options.pwprompt) {
+			if (lookup_password() == 1) {
+				return(EXIT_FAILURE);
+			}
```
Comment (andy-bower) [inline src/rc.c:202]: This means we can't fall back to prompt, .netrc or gpg if the password hasn't been cached?

**Review by andy-bower — COMMENTED** [review]
Thanks for enhancing my PR @t-brown! I think it was a good idea to split out the password prompting and libsecret integration.

I have some questions about the decision tree for using the various methods.

Admittedly these were open questions when I raised the PR but I just want to check this is how you'd like it!

In particular perhaps we could think of the first time user who potentially could run without a config file if they put the URL on the command line and their user ID matched the system one?

Anyway, I think the choices you've come up with are the conservative ones which would avoid surprising behaviour for existing users and there'd be scope to change.

File: src/secret.c (hunk 4 above)
Comment (andy-bower) [reply src/secret.c:52]: Fair enough! I don't think there's a particular need not just to put the full identifier in one string constant - do you want to do that change?

File: src/rc.c (hunk 6 above)
Comment (t-brown) [reply src/rc.c:202]: Yeah, I'm thinking these are not fall backs, rather "or"s.  In my head the workflow would be to run `-p` the first time and then `libsecret` afterwards. Or you just use `.netrc` or `gpg`.  But not to prompt the user, since that is an CLI option.

File: src/rc.c (hunk 5 above)
Comment (t-brown) [reply src/rc.c:195]: I was thinking of this logic, where if the option to prompt is selected, we just use that no matter what. Rather than falling back if it's not found in any places (`netrc`, `gpg`, `libsecret`).

File: src/rc.c (hunk 6 above)
Comment (andy-bower) [reply src/rc.c:202]: Re this and last comment - I think you're unambiguously right about **prompt vs .netrc/gpg** but I'm not so sure when it comes to **prompt vs. libsecret** as I think those two modes would typically go hand-in-hand with other applications.

But I *think* I get more or less what I want if `libsecret = yes` is specified in the config as then you can simply add `-p` if you need to specify what that password is.

What about a command line option (e.g. `-k`- see below!) as a compromise to enable this config to make it easier for a first-time user to play?

File: src/rc.c [hunk 7]
```diff
@@ -147,6 +149,12 @@ read_rc(const char *file)
 				} else {
 					options.netrc = 0;
 				}
+			} else if (strncmp("libsecret", vals[0], 9) == 0) {
```
Comment (andy-bower) [inline src/rc.c:152]: A thought: how about "keyring" or similar as the user-facing name for this config option as `libsecret` is more of an implementation detail and the user is more likely to be used to prompts from the likes of 'gnome-keyring' or 'kwallet' that actually implement the storage?

Comment (andy-bower) [reply src/rc.c:152]: (we'd better add to the man page too @t-brown!)

Comment (t-brown) [reply src/rc.c:152]: I do like the word `keyring` better, however I am torn as we are using `libsecret`.  Does `kwallet` use `libsecret`? As it's GNOME vs KDE?

In hind sight, I should have done something like the following in the `rc` file:
```
pass = netrc | gpg | keyring 
```

File: src/secret.c (hunk 4 above)
Comment (andy-bower) [reply src/secret.c:52]: Also, you may have a better schema id in mind - I just constructed one from this repo by default!

File: src/rc.c (hunk 7 above)
Comment (andy-bower) [reply src/rc.c:152]: To be honest, I'm not sure how the potential backends are organised and could have been wrong to mention kwallet - I think the actual API name might be "Secrets Service". One could also perhaps generically call it a credentials cache.
